File: onnxsim/memory_planning.cpp
#include "memory_planning.h"

#include <algorithm>
#include <set>

namespace onnxsim {

namespace {

struct Interval {
  std::string name;
  int64_t size = 0;
  int start =
      0;  // node index after which the tensor is available (-1 = before node 0)
  int end = 0;  // last node index at which the tensor is still needed
};

// True when `a` and `b` are never simultaneously live, i.e. safe to share
// offset space. Conservative at the boundary: a tensor whose last use is node
// i and one produced by node i itself still count as overlapping (`a.end ==
// b.start` is not `<`), since a node's own output is not guaranteed safe to
// alias with its inputs' storage in general -- that exception is handled
// separately and explicitly by the in-place-aliasing pass below, for the
// specific ops it is actually safe for.
bool Disjoint(const Interval& a, const Interval& b) {
  return a.end < b.start || b.end < a.start;
}

// Ops whose ONNX semantics guarantee a single output that is strictly
// shape- and dtype-identical to their first input, computed purely
// elementwise (output[i] depends only on input[i]) -- so an executor can
// always compute the output by overwriting the input's own storage in
// place, with no numerical difference from writing to a separate buffer.
// This is a curated allowlist, not a schema query (nothing here parses
// attributes or checks the actual input/output types), so it deliberately
// excludes anything with an exception -- e.g. not "Transpose", whose output
// (unless the permutation happens to be the identity, which nothing here
// checks) physically reorders elements rather than computing the same
// index in place.
bool IsInPlaceSafeOp(const std::string& op_type) {
  static const std::set<std::string> kSafeOps = {
      "Relu",     "Sigmoid",  "Tanh",        "LeakyRelu",  "Elu",      "Selu",
      "Softplus", "Softsign", "HardSigmoid", "Clip",       "Neg",      "Abs",
      "Sqrt",     "Exp",      "Log",         "Reciprocal", "Identity", "Erf",
      "Celu",     "Round",    "Ceil",        "Floor",      "Sign",
  };
  return kSafeOps.count(op_type) > 0;
}

// Ops that reinterpret their first input's element sequence under a
// different shape without moving any bytes -- ONNX requires row-major/
// C-contiguous tensor layout, and none of these ops permute element order
// (they only regroup, split, or insert/remove size-1 dimensions), so their
// output is byte-for-byte identical to their input whenever the sizes
// TensorBytes() reports actually agree (checked the same as every other
// candidate, at the call site). This is a *stronger* guarantee than
// IsInPlaceSafeOp's "safe to compute in place" -- these ops need no
// computation at all, the output is a pure view -- but it is folded into
// the same union-find aliasing pass since the resulting placement is
// identical: one shared slot for input and output alike. Not "Transpose"
// (see IsInPlaceSafeOp) or "Expand"/"Tile" (which do move/duplicate bytes).
bool IsViewOp(const std::string& op_type) {
  static const std::set<std::string> kViewOps = {
      "Reshape",
      "Flatten",
      "Squeeze",
      "Unsqueeze",
  };
  return kViewOps.count(op_type) > 0;
}

// Union-find (disjoint-set) over tensor names, path-compressed, used to
// coalesce a chain of in-place-eligible ops (e.g. Relu -> Sigmoid -> Tanh)
// into one placement group that needs a single slot for its whole span
// rather than one slot per tensor. Passes/returns names by value (plain
// std::string copies) rather than references into `parent_`, since Find()
// mutates that same map recursively -- returning a reference into a std::map
// entry that a nested call may go on to overwrite is the kind of aliasing
// hazard worth just not risking for a data structure this small.
class UnionFind {
 public:
  std::string Find(const std::string& name) {
    auto it = parent_.find(name);
    if (it == parent_.end()) {
      parent_[name] = name;
      return name;
    }
    if (it->second == name) return name;
    const std::string root = Find(it->second);
    parent_[name] = root;  // path compression
    return root;
  }

  void Union(const std::string& a, const std::string& b) {
    const std::string ra = Find(a);
    const std::string rb = Find(b);
    if (ra != rb) parent_[ra] = rb;
  }

 private:
  std::map<std::string, std::string> parent_;
};

}  // namespace

MemoryPlan ComputeActivationMemoryPlan(const GraphView& graph) {
  MemoryPlan plan;

  const std::set<std::string> weights(graph.initializers.begin(),
                                      graph.initializers.end());
  const std::set<std::string> graph_inputs(graph.inputs.begin(),
                                           graph.inputs.end());
  const std::set<std::string> graph_outputs(graph.outputs.begin(),
                                            graph.outputs.end());

  // Last node index consuming each tensor; a graph output is "consumed" at
  // `end` so it stays live through the whole pass -- the same liveness
  // convention PeakMemoryFootprint uses (model_metrics.cpp).
  const int end = static_cast<int>(graph.nodes.size());
  std::map<std::string, int> last_use;
  for (int i = 0; i < end; ++i) {
    for (const std::string& name : graph.nodes[i].inputs) {
      if (!name.empty()) last_use[name] = i;
    }
  }
  for (const std::string& out : graph.outputs) last_use[out] = end;

  // Producer node index of each non-weight tensor: -1 for a graph input
  // (available before node 0), else the index of the node that outputs it.
  std::map<std::string, int> produced_at;
  for (const std::string& in : graph.inputs) {
    if (weights.find(in) == weights.end()) produced_at[in] = -1;
  }
  for (int i = 0; i < end; ++i) {
    for (const std::string& name : graph.nodes[i].outputs) {
      if (!name.empty() && weights.find(name) == weights.end()) {
        produced_at[name] = i;
      }
    }
  }

  // How many (node, input-slot) pairs consume each tensor across the whole
  // graph -- the in-place-aliasing pass below only aliases a tensor whose
  // sole consumer is the one candidate node, so overwriting it can never
  // corrupt a read some other node still needs.
  std::map<std::string, int> consumer_count;
  for (const NodeView& node : graph.nodes) {
    for (const std::string& name : node.inputs) {
      if (!name.empty()) ++consumer_count[name];
    }
  }

  // In-place aliasing: union a safe unary/view op's input[0] with its
  // output[0] (see IsInPlaceSafeOp and IsViewOp) whenever doing so is
  // provably safe --
  //   * not a weight, a graph input, or a declared graph output (an
  //     externally-owned buffer, or one the caller reads only after the
  //     whole graph finishes, so overwriting it mid-run would be observed
  //     as corruption rather than reused space);
  //   * consumed exactly once, by this node -- the only read of the
  //     original value;
  //   * input and output agree on size (always true for these ops on a
  //     well-formed model; checked anyway as a defensive belt-and-braces).
  // A chain of such ops unions transitively into one group -- Find() on any
  // member returns the same root -- that needs only a single slot for its
  // entire span, rather than one slot per node. The chain may freely mix
  // both categories (e.g. Reshape -> Relu -> Flatten): the eligibility
  // conditions and the resulting placement are identical either way.
  UnionFind uf;
  for (const NodeView& node : graph.nodes) {
    if (!IsInPlaceSafeOp(node.op_type) && !IsViewOp(node.op_type)) continue;
    if (node.inputs.empty() || node.outputs.size() != 1) continue;
    const std::string& in0 = node.inputs[0];
    const std::string& out0 = node.outputs[0];
    if (in0.empty() || out0.empty()) continue;
    if (weights.count(in0) || graph_inputs.count(in0) ||
        graph_outputs.count(in0)) {
      continue;
    }
    const auto cit = consumer_count.find(in0);
    if (cit == consumer_count.end() || cit->second != 1) continue;

    const auto in_bytes = TensorBytes(in0, graph.shapes, graph.dtypes);
    const auto out_bytes = TensorBytes(out0, graph.shapes, graph.dtypes);
    if (!in_bytes || !out_bytes || in_bytes->is_symbolic() ||
        out_bytes->is_symbolic()) {
      continue;
    }
    if (in_bytes->to_int() != out_bytes->to_int()) continue;

    uf.Union(in0, out0);
  }

  // Pass 1: one Interval per plannable *group* -- a group is a single
  // tensor, or several coalesced by the aliasing pass above -- with
  // everything unplannable going to `unplanned`. An unconsumed intermediate
  // (no `last_use` entry, not a graph output) defaults to staying live
  // through `end`, the same "never freed" fallback the live-set walk behind
  // PeakMemoryFootprint implicitly applies to it. `naive_bytes` is summed
  // per tensor regardless of grouping: it is the no-optimization-at-all
  // baseline, not affected by how well this planner happens to pack things.
  struct GroupSpan {
    int64_t size = 0;
    int start = 0;
    int end = 0;
    bool seen = false;
  };
  std::map<std::string, GroupSpan> groups;  // keyed by union-find root

  for (const auto& [name, start] : produced_at) {
    const auto bytes = TensorBytes(name, graph.shapes, graph.dtypes);
    if (!bytes || bytes->is_symbolic()) {
      plan.unplanned.push_back(name);
      continue;
    }
    const auto it = last_use.find(name);
    const int last = it == last_use.end() ? end : it->second;
    plan.naive_bytes += bytes->to_int();

    GroupSpan& g = groups[uf.Find(name)];
    if (!g.seen) {
      g = {bytes->to_int(), start, last, true};
    } else {
      g.start = std::min(g.start, start);
      g.end = std::max(g.end, last);
      g.size = std::max(g.size, bytes->to_int());
    }
  }

  std::vector<Interval> intervals;
  intervals.reserve(groups.size());
  for (const auto& [root, g] : groups) {
    intervals.push_back({root, g.size, g.start, g.end});
  }

  // Pass 2: greedy best-fit placement, largest group first (ties broken by
  // root name for determinism) -- the standard linear-scan/greedy-by-size
  // register allocation heuristic, which is what makes this
  // "fragmentation-aware" rather than a naive bump allocator. For each
  // group, gather the offset ranges of every already-placed group whose
  // liveness overlaps it (the ranges it must avoid), then take the smallest
  // gap between/around them that fits, falling back to appending after the
  // last blocker.
  std::sort(intervals.begin(), intervals.end(),
            [](const Interval& a, const Interval& b) {
              if (a.size != b.size) return a.size > b.size;
              return a.name < b.name;
            });

  std::vector<Interval> placed;
  std::vector<std::pair<int64_t, int64_t>>
      placed_ranges;  // parallel: [off, off+size)
  std::map<std::string, std::pair<int64_t, int64_t>> group_offsets;

  for (const Interval& iv : intervals) {
    std::vector<std::pair<int64_t, int64_t>> blocking;
    for (std::size_t i = 0; i < placed.size(); ++i) {
      if (!Disjoint(iv, placed[i])) blocking.push_back(placed_ranges[i]);
    }
    std::sort(blocking.begin(), blocking.end());

    int64_t best_offset = -1;
    int64_t best_gap = -1;
    int64_t cursor = 0;
    for (const auto& [b_start, b_end] : blocking) {
      const int64_t gap = b_start - cursor;
      if (gap >= iv.size && (best_gap < 0 || gap < best_gap)) {
        best_offset = cursor;
        best_gap = gap;
      }
      cursor = std::max(cursor, b_end);
    }
    if (best_offset < 0)
      best_offset = cursor;  // no gap fit; append after every blocker

    group_offsets[iv.name] = {best_offset, iv.size};
    placed.push_back(iv);
    placed_ranges.push_back({best_offset, best_offset + iv.size});
    plan.arena_bytes = std::max(plan.arena_bytes, best_offset + iv.size);
  }

  // Expand each group's single placement to every tensor that was coalesced
  // into it -- an aliased tensor and the value(s) it shares storage with all
  // report the identical (offset, size).
  for (const auto& [name, start] : produced_at) {
    (void)start;
    const auto it = group_offsets.find(uf.Find(name));
    if (it != group_offsets.end()) plan.offsets[name] = it->second;
  }

  return plan;
}

}  // namespace onnxsim

File: onnxsim/memory_planning.py
import copy
import dataclasses
from typing import Dict, List, Tuple

import onnx
from rich import print
from rich.table import Table
from rich.text import Text

from onnxsim.model_info import (
    METADATA_PREFIX,
    ModelInfo,
    _set_metadata,
    human_readable_size,
)

__all__ = [
    "MemoryPlan",
    "plan_activation_memory",
    "print_memory_plan",
    "annotate_memory_plan",
]


@dataclasses.dataclass(frozen=True)
class MemoryPlan:
    """A static byte-offset allocation plan for a model's activation tensors,
    computed by :func:`plan_activation_memory`.

    Every *planned* tensor (a graph input, node output, or graph output with a
    concretely known size) gets a ``(offset, size)`` pair in
    :attr:`tensor_offsets`, all within one shared arena of
    :attr:`arena_bytes`: two tensors only ever get overlapping offset ranges
    when their liveness intervals do not overlap, so a deployment target can
    allocate a single :attr:`arena_bytes`-sized buffer and hand each tensor
    its own slice, reused over the tensor's lifetime, instead of one
    permanent buffer per activation (:attr:`naive_bytes` -- the baseline
    :attr:`compression_ratio` is measured against).

    Weights (initializers) are never part of the plan: they stay resident for
    the whole graph, outside this arena. A tensor whose size cannot be
    resolved to a concrete byte count -- unknown shape/dtype, or a dynamic
    dimension -- is excluded from both :attr:`tensor_offsets` and
    :attr:`naive_bytes` and listed in :attr:`unplanned` instead of guessed, so
    a plan with a non-empty :attr:`unplanned` is a partial lower bound, not a
    complete plan.

    Only the top-level graph is planned -- tensors inside control-flow
    (``If``/``Loop``/``Scan``) subgraph bodies are not visited at all.
    """

    arena_bytes: int
    naive_bytes: int
    tensor_offsets: Dict[str, Tuple[int, int]]
    unplanned: List[str]

    @property
    def compression_ratio(self) -> float:
        """Fraction of :attr:`naive_bytes` the plan saves:
        ``1 - arena_bytes / naive_bytes``. ``0.0`` when there is nothing
        plannable (``naive_bytes == 0``).
        """
        if self.naive_bytes == 0:
            return 0.0
        return 1.0 - self.arena_bytes / self.naive_bytes


def plan_activation_memory(
    model: onnx.ModelProto, run_shape_inference: bool = True
) -> MemoryPlan:
    """Compute a :class:`MemoryPlan` for ``model``'s top-level graph.

    Delegates the allocation itself to the C++ core
    (``onnxsim/memory_planning.h``): a greedy best-fit placement over each
    tensor's liveness interval (from production to last use, the same
    liveness convention :class:`onnxsim.model_info.ModelInfo`'s
    ``memory_footprint`` uses), largest tensor first. ``model`` is not
    modified. Pass ``run_shape_inference=False`` when ``model`` already
    carries populated shapes (e.g. inferred with data propagation) to skip
    the extra pass.
    """
    from onnxsim import onnxsim_cpp2py_export as _C

    offsets, arena_bytes, naive_bytes, unplanned = _C._memory_plan(
        model.SerializeToString(), run_shape_inference
    )
    return MemoryPlan(
        arena_bytes=arena_bytes,
        naive_bytes=naive_bytes,
        tensor_offsets=dict(offsets),
        unplanned=list(unplanned),
    )


def print_memory_plan(plan: MemoryPlan, limit: int = 50) -> None:
    """Pretty-print ``plan`` as a table of tensor offsets (ordered by offset,
    then name), followed by the arena/naive totals and compression ratio, and
    -- when non-empty -- the list of tensors :func:`plan_activation_memory`
    could not place. Capped at ``limit`` rows (with a "... and N more" line)
    so a large model's plan stays readable.
    """
    table = Table(title="Activation Memory Plan")
    table.add_column("Tensor")
    table.add_column("Offset")
    table.add_column("Size")

    ordered = sorted(plan.tensor_offsets.items(), key=lambda kv: (kv[1][0], kv[0]))
    for name, (offset, size) in ordered[:limit]:
        table.add_row(name, human_readable_size(offset), human_readable_size(size))
    print(table)
    if len(ordered) > limit:
        print(Text(f"... and {len(ordered) - limit} more", style="dim"))

    print(
        f"Arena: {human_readable_size(plan.arena_bytes)} "
        f"(naive: {human_readable_size(plan.naive_bytes)}, "
        f"{plan.compression_ratio * 100:.1f}% smaller)"
    )
    if plan.unplanned:
        shown = ", ".join(plan.unplanned[:10])
        more = ", ..." if len(plan.unplanned) > 10 else ""
        print(
            Text(
                f"{len(plan.unplanned)} tensor(s) could not be planned "
                f"(unknown shape/dtype or a dynamic dimension): {shown}{more}",
                style="yellow",
            )
        )


def _join_capped(items: List[str], limit: int) -> str:
    # Same "capped list, one line" shape as model_info.h's C++ JoinCapped /
    # RecordCappedListMetadata: join up to `limit` entries with ", ", appending
    # a "... (+N more)" marker when there were more, so a long unplanned list
    # doesn't blow up a single metadata_props value.
    joined = ", ".join(items[:limit])
    if len(items) > limit:
        joined += f", ... (+{len(items) - limit} more)"
    return joined


def annotate_memory_plan(
    model: onnx.ModelProto, prefix: str = METADATA_PREFIX
) -> onnx.ModelProto:
    """Return a shape-inferred copy of ``model`` with :func:`plan_activation_memory`'s
    result stored in ``metadata_props``, so a downstream consumer -- an embedded
    runtime, a code generator, anything that doesn't have onnxsim installed --
    can read the plan directly off the model instead of recomputing it:

    - **Model**: ``<prefix>memory_plan_arena_bytes``, ``naive_bytes`` and
      ``compression_ratio`` (:class:`MemoryPlan`'s totals); ``unplanned_count``
      always, plus ``unplanned`` (a capped, comma-joined list of names) when
      non-empty.
    - **Value** (inputs, outputs, value_info -- i.e. every *planned* tensor):
      ``<prefix>mem_offset`` and ``<prefix>mem_size``, matching
      :attr:`MemoryPlan.tensor_offsets`. A tensor onnxsim could not plan (see
      :attr:`MemoryPlan.unplanned`) is simply left unannotated, the same way
      :func:`annotate_metadata` leaves an unknown-shape tensor's ``bytes``
      unset rather than guessing.

    Values are strings. The input model is never mutated; the returned copy is
    shape-inferred so intermediate values carry a matching ``value_info`` entry
    to annotate. Weights are never annotated here -- they are outside the
    activation arena by construction (see :func:`plan_activation_memory`).
    """
    plan = plan_activation_memory(model)
    # Work on an inferred copy: infer_shapes returns a fresh model on success,
    # but the original object on failure -- deep-copy first so the caller's
    # model is never touched and the annotations land on populated value_info.
    # Shape inference runs a second time here (plan_activation_memory already
    # ran it once internally, in C++, to compute the plan itself) -- the same
    # duplication onnxsim.model_info.annotate_metadata already accepts, for
    # the same reason: the C++ side never hands its shape-inferred model back.
    work = ModelInfo._infer_shapes(copy.deepcopy(model))

    value_infos = {
        vi.name: vi
        for vi in list(work.graph.input)
        + list(work.graph.output)
        + list(work.graph.value_info)
    }
    for name, (offset, size) in plan.tensor_offsets.items():
        vi = value_infos.get(name)
        if vi is not None:
            _set_metadata(vi, prefix + "mem_offset", str(offset))
            _set_metadata(vi, prefix + "mem_size", str(size))

    _set_metadata(work, prefix + "memory_plan_arena_bytes", str(plan.arena_bytes))
    _set_metadata(work, prefix + "memory_plan_naive_bytes", str(plan.naive_bytes))
    _set_metadata(
        work,
        prefix + "memory_plan_compression_ratio",
        f"{plan.compression_ratio:.4f}",
    )
    _set_metadata(
        work, prefix + "memory_plan_unplanned_count", str(len(plan.unplanned))
    )
    if plan.unplanned:
        _set_metadata(
            work, prefix + "memory_plan_unplanned", _join_capped(plan.unplanned, 20)
        )
    return work
